Generic support of influxdb format

dpino opened the pull request:
Supersedes #1037.

To recap, an influxdb entry meets the following formatting:

```
ULEAF(,K=V)* value=VALUE
```

Where ULEAF in an unique leaf in the schema. If the leaf is not unique, ULEAF is the absolute path to the leaf. The same idea is applied to keys of tags.

Leaves and keys uniqueness is a property of the schema, not the data.

Here's a WIP implementation. So far, the influxdb printer inherits the same structure as YANG and XML printers. The main difference is decorating grammar nodes with an extra function that tells whether the node is duplicated in the schema or not. Initially I wanted the method to return the full path for the node, but this is not a good idea cause due to the use of `group` and `uses` keywords in YANG there are nodes that are single instances in the schema. If I decorate the node with a function that returns the full path value, the value will be the same independently of what route was taken to reach the node. 

Example:

```
$ sudo ./snabb lwaftr run --name lwaftr --conf lwaftr.conf --on-a-stick 82:00.0
```

```
$ sudo ./snabb config get-state --schema=snabb-softwire-v2 --format=influxdb lwaftr /
softwire-config/instance/softwire-state/drop-all-ipv4-iface-bytes,device=82:00.0 value=0
softwire-config/instance/softwire-state/drop-all-ipv4-iface-packets,device=82:00.0 value=0
softwire-config/instance/softwire-state/drop-all-ipv6-iface-bytes,device=82:00.0 value=0
softwire-config/instance/softwire-state/drop-all-ipv6-iface-packets,device=82:00.0 value=0
softwire-state/drop-all-ipv4-iface-bytes value=0
softwire-state/drop-all-ipv4-iface-packets value=0
softwire-state/drop-all-ipv6-iface-bytes value=0
softwire-state/drop-all-ipv6-iface-packets value=0
```

Uniqueness of keys is pending.

dpino committed:
Add basic support for influxdb format

dpino committed:
Prevent potential error if value is nil

wingo commented on src/lib/yang/data.lua:
```diff
@@ -222,14 +222,43 @@ function data_grammar_from_schema(schema, is_config)
+            if v.kind == 'leaf' then
+               if keys[k] then duplicates[k] = true end
+               keys[k] = true
+               v.duplicated = function()
```
Why does this need to be a function if it just returns a constant?  Could we annotate the grammar with the "unique key" in all cases?  Since the grammar is only as large as the schema, not the data, it's no big deal to add this.

If we keep this "add .duplicated property" pass, and the property value is a function of the schema and not the data, then the computation should be done in schema.lua -- the schema is read-only from this function's perspective.

wingo reviewed (commented):
Needs tests.  Besides that, some comments also.

dpino committed:
Initialize root variable if nil

dpino committed:
Add influxdb printer tests

dpino commented:
Added several improvements. It still needs more tests though. Example of querying config, instead of state:

```
softwire-config/binding-table/softwire/b4-ipv6,ipv4=193.5.1.100,psid=5 value=fc00:1:2:3:4:5:0:82
softwire-config/binding-table/softwire/br-address,ipv4=193.5.1.100,psid=5 value=fc00::100
softwire-config/binding-table/softwire/port-set/psid-length,ipv4=193.5.1.100,psid=5 value=6
softwire-config/binding-table/softwire/b4-ipv6,ipv4=193.5.1.100,psid=53 value=fc00:1:2:3:4:5:0:b2
...
softwire-config/external-interface/allow-incoming-icmp value=true
softwire-config/external-interface/error-rate-limiting/packets value=600000
softwire-config/external-interface/reassembly/max-fragments-per-packet value=40
softwire-config/instance/queue/external-interface/ip,id=1 value=10.0.1.1
softwire-config/instance/queue/external-interface/mac,id=1 value=02:aa:aa:aa:aa:aa
softwire-config/instance/queue/external-interface/next-hop/mac,id=1 value=02:99:99:99:99:99
softwire-config/instance/queue/internal-interface/ip,id=1 value=fc00::100
softwire-config/instance/queue/internal-interface/mac,id=1 value=02:aa:aa:aa:aa:aa
softwire-config/instance/queue/internal-interface/next-hop/mac,id=1 value=02:99:99:99:99:99
softwire-config/internal-interface/allow-incoming-icmp value=true
softwire-config/internal-interface/error-rate-limiting/packets value=600000
softwire-config/internal-interface/reassembly/max-fragments-per-packet value=40
softwire-config/name value="lwaftr"
```

dpino commented:
Added a few more test cases and fixed several errors in serialization to influxdb format. I can add more test cases, suggestions welcome.

dpino committed:
Disambiguate keys

dpino commented:
Added commit that disambiguates keys. In case two or more keys occurred in the schema, the key (tag) is an absolute path.

dpino commented:
I'm not sure what to about arrays. The tag name is position. Should the tag name be an absolute path if there are two or more arrays (leaf-list) in an schema?

wingo commented:
Because leaf-lists are leaves, you won't have two position()=x in a key.  It's sufficient to uniquify the name of the leaf I think.  However I think it could be a good idea to use a different name other than "position", otherwise you might have a conflict with a "position=x" where that position is a key of a list (not a leaf-list).  So use e.g. %position or %pos or something that isn't a valid YANG identifier as the tag maybe?

dpino committed:
Use '%position' as tag key for leaf-list nodes

dpino committed:
Detect nodes uniqueness as a function of the schema

A helper class Marker is used to mark whether nodes are unique or
not. When 'uses' nodes are expanded, if they're repeated then their
hanging leaves nodes are marked as non unique.

Also created method `print_entry` that is used to print scalar nodes
and array nodes.  In the case of array nodes, position is prepend
with the full node path is node is not unique.

dpino committed:
Escape influxdb tags and values

dpino committed:
Fix several errors in serialization of arrays

dpino commented:
Fixed position serialization as suggested. Also rebased PR and squashed several commits.

wingo commented:
I am not sure we are getting it right as regards strings (see https://github.com/Igalia/snabb/pull/1037#issuecomment-375296119).  Can you verify?

dpino commented:
Np, I can add a new test case.

dpino committed:
Schema markers need to be reset on each load

dpino committed:
Test more data types in test case

wingo commented on src/lib/yang/data.lua:
```diff
@@ -1012,6 +1013,270 @@ function xpath_printer_from_grammar(production, print_default, root)
+         return '"'..escape_double_quotes(value)..'"'
+      elseif primitive_type == 'boolean' then
+         return value and 'true' or 'false'
+      else
```
Don't you have to add "i" as a suffix for integer types?  Likewise I think `tostring` is sufficient for decimal64, no?

wingo commented on src/lib/yang/data.lua:
```diff
@@ -1245,6 +1510,142 @@ function rpc_output_printer_from_schema(schema)
+            bar {baz "baz"; y "y"; z "z";}
+         }
+      ]], [[
+         foo/bar/y,baz=baz value="y"
```
For the uniquified names, WDYT about adding an initial / prefix?  That would be a nice visual indicator that we're dealing with an absolute path.

wingo commented on src/lib/yang/schema.lua:
```diff
@@ -644,6 +671,8 @@ end
 -- Resolve all augment nodes. (TODO)
 function resolve(schema, features)
+   marker.leaf:reset()
+   marker.uses:reset()
```
I would rather add the uniqueness detection as a post-pass, once we've parsed everything.  Like the pass that computes whether a node is config or state.  That would avoid the global marker set; the state can be specific to the pass.  WDYT?

dpino replied:
OK, I was not aware there was already an extra pass. I can add the computation of uniqueness to that pass. Yep, I think in that case the marker won't be necessary as uniqueness state goes appended to each node.

wingo reviewed (commented):
looking good, just three comments

dpino committed:
Handle integer data types

Small fix in handle of float values: Use tostring(float).

dpino committed:
Append '/' to uniquified names

dpino committed:
Collect node uniqueness in extra pass

wingo commented on src/lib/yang/schema.lua:
```diff
@@ -925,13 +926,36 @@ function parse_schema_file(filename)
+   end
+   visit(s)
+   return s
+end
```
ah, very nice.

wingo reviewed (approved):
lgtm when ci is green.  thanks for putting up with my pickiness!!!
